added google sign in and code attribution

christenmartin opened the pull request.

christenmartin committed:
added google oauth

christenmartin committed:
added google oauth sign-in

christenmartin committed:
added google sign in

christenmartin committed:
Merge remote-tracking branch 'upstream/master'
quit
\q
git merge--abort;

git merge-- abort

git merge- abort;

ZZZZZZZZ
ZZ
I
i
i

christenmartin committed:
fixing routes

awilson28 commented on server/passport-setup.js:
```diff
@@ -0,0 +1,27 @@
+  },
+  function(accessToken, refreshToken, profile, done) {
+       User.findOrCreate({ googleId: profile.id, username: profile.displayName, email: profile._json.email }, function (err, user) {
+         return done(err, user);
```
this is definitely a nit: you can pass done as the callback : 
` User.findOrCreate({ googleId: profile.id, username: profile.displayName, email: profile._json.email }, done)

christenmartin committed:
changed generateSlots route to cron job

christenmartin committed:
fixing cron

christenmartin committed:
Merge branch 'slotsWork'

christenmartin committed:
changed callback per suggestion

christenmartin commented:
-Migrated generateSlots logic to cron job instead of route.
-Implemented Ayana’s customRoutes suggestions.
-Added nodemailer and /api/mail post route.

awilson28 commented on server/boot/customRoutes.js:
```diff
@@ -4,75 +4,105 @@ module.exports = function (app) {
       .then(slots => {
-        if (!slots) {
-          res.json({ "error": "No slots created for this date and storeId yet."})
+        if (slots.length < 1) {
```
you can also write this as `if(!slots.length){...`

awilson28 commented on server/boot/customRoutes.js:
```diff
@@ -4,75 +4,105 @@ module.exports = function (app) {
+          return Slot.create(newSlots);
+        })
+      })
+      .then(createdSlots => createdSlots)
```
nit: you can remove this line for two reasons - 1) retrieving the resolved value of a promise solely to return the resolved value is unnecessary/redundant and 2) the anonymous fn passed to `cron.schedule` isn't interested in `createdSlots` - the only purpose is to create the slots. If the anonymous fn were concerned with the results, you'd want to use `Promise.all` instead of the synchronous Array `map` method to handle the promises. Here's a demonstration of the difference: https://repl.it/repls/FelineThirdCryptos

awilson28 commented on server/passport-setup.js:
```diff
@@ -0,0 +1,25 @@
+});
+
+passport.deserializeUser((id, done) => {
+  User.findById(id).then((user) => {
```
a nit for sure: why not stick with error-first callbacks instead of switching to promises here? just to adhere to the other patterns in this file.

awilson28 commented on server/boot/customRoutes.js:
```diff
@@ -4,75 +4,105 @@ module.exports = function (app) {
 
+  
+  //Cron job-- triggers each day at 1am GMT to generate the next day's slots for each store.
+  cron.schedule("0 0 * * *", function() {
```
it probably makes sense to put this cron job in a different file for organizational reasons and then require that file here or in the main app.js

christenmartin replied:
this might be a dumb question, but how would I call/require it here or in main? I considered putting the cron scheduler in a separate file in a fn like initCron() and then calling it from boot/main, but then the enclosing fn would only run once when called. Are you suggesting something like put the query logic in a separate fn in another file, and then use cron.schedule(cron expression, fnFromOtherFile()) in customRoutes?

This might end up being a moot point since I think we're going to switch over to dynamoDB like you suggested, and then use an AWS lambda function to trigger the cron job. (Or Antanas suggested possibly heroku cloud db). But still curious to know the answer.

awilson28 replied:
There are no dumb questions, especially since my main value add right now is answering questions :). I think your question is really about how cron jobs are queued and managed by the event loop. When the event loop encounters a setInterval/cron job, it enqueues the task as well as the callback to be executed and will continue to execute the callback at the given interval even if there are no other server tasks. Put differently, setIntervals/cron jobs keep the event loop full in the same way that `server.listen()` keeps the event loop full. In fact, `server.listen()` uses `process.nextTick` (https://medium.com/@amanhimself/how-process-nexttick-works-in-node-js-cb327812e083) under the hood. Read more about it here: https://nodejs.org/fa/docs/guides/event-loop-timers-and-nexttick/

Here are some examples that I think will clarify how cron jobs are managed; try them out and see: 

// // Example 1 - everything in the main server file (index.js)
// // to run: node index.js
const express = require('express');
const cron = require('node-cron');
app = express();
cron.schedule("* * * * * *", function() {
  console.log("running a task every second");
});
app.listen(3128);

-------------------------------

// // Example 2 - main server file requires cron-job.js 
// // to run: node index.js
// server file: 
const express = require('express');
app = express();
require('./cron-job.js');
app.listen(3128);

// cron-job.js
const cron = require('node-cron');
cron.schedule("* * * * * *", function() {
  console.log("running a task every second");
});

-------------------------------------

// // Example 3 - running the cron job without running the server file 
// // to run: node cron-job.js
// cron-job.js
const cron = require('node-cron');
cron.schedule("* * * * * *", function() {
  console.log("running a task every second");
});

--------------------------------------

So I was suggesting that we put the cron job in a separate file as demonstrated in Example 2.

awilson28 replied:
I'm all for moving toward DynamoDb and Lambda, which is perfect for handling cron jobs. And DynamoDb will simplify database management and development, since we won't have to manange local databases. Too, these are great technologies to know before you enter the job market. Let me know if you want to discuss implementation.

christenmartin replied:
Thanks for the depth of the answer. For example 2, how come we need to require node-cron in both server.js and cron-job.js? Wouldn't it be sufficient to only require cron-job.js in server, since it already contains node-cron? 

Re: talking implementation, that would be great. I haven’t looked deeply into the implementation details yet, but I’ll do so soon. Are there some implementation decisions you’d suggest we start thinking about beforehand?

awilson28 replied:
In example 2, I left `node-cron` in server.js accidentally, you're right that it's not necessary - only cron-job.js needs `node-cron`

awilson28 replied:
@christenmartin The main thing is to understand how DynamoDB handles partition keys and sort keys. Basically, you can use a single DynamoDB table as a database, because records are identified by partition key first, then sort key. 
https://aws.amazon.com/blogs/database/choosing-the-right-dynamodb-partition-key/

christenmartin committed:
added nodemailer with route

christenmartin committed:
added store name to post mail route

awilson28 commented on package.json:
```diff
@@ -12,14 +12,20 @@
     "dev": "nodemon ."
   },
   "dependencies": {
+    "ajv": "^5.5.2",
```
this doesn't appear to be used

awilson28 commented on server/boot/customRoutes.js:
```diff
@@ -4,75 +4,135 @@ module.exports = function (app) {
+     if (err) {
+       console.log(err)
+     } else {
+       console.log(info)
```
since you're logging both the error and info, you can just do `console.log(err, info)`

christenmartin committed:
hopefully fixed nodemailer issue

christenmartin committed:
changed availSlots route to allSlotsToday route

christenmartin committed:
superfluous comment for heroku rebuild

christenmartin committed:
removed comment

christenmartin committed:
added moment dependency

christenmartin committed:
frustration

christenmartin committed:
moved moment from dev dependency to regular dependency

christenmartin committed:
added keys prod vs dev for heroku

christenmartin committed:
mongo_URI

christenmartin committed:
miniscule change
